presto-main/src/main/java/com/facebook/presto/dispatcher/DispatchManager.java

```diff
@@ -0,0 +1,323 @@
+            DispatchQuery failedDispatchQuery = failedDispatchQueryFactory.createFailedDispatchQuery(
+                    session,
+                    query,
+                    Optional.ofNullable(selectionContext).map(SelectionContext::getResourceGroupId),
+                    e);
+
+            try {
+                queryCreated(failedDispatchQuery);
+            }
+            finally {
+                handleQueryFailure(failedDispatchQuery);
+            }
+            return;
+        }
+
+        try {
+            queryCreated(dispatchQuery);
+            dispatchQuery.addStateChangeListener(newState -> {
+                if (newState.isDone()) {
+                    stats.queryFinished(dispatchQuery.getBasicQueryInfo());
+                }
+            });
+
+            resourceGroupManager.submit(preparedQuery.getStatement(), dispatchQuery, selectionContext, queryManagementExecutor);
+        }
+        catch (RuntimeException e) {
+            dispatchQuery.fail(e);
+        }
+    }
+
+    private void queryCreated(DispatchQuery dispatchQuery)
+    {
+        queryTracker.addQuery(dispatchQuery);
+        stats.queryQueued();
+    }
+
+    private void handleQueryFailure(DispatchQuery dispatchQuery)
+    {
+        try {
+            stats.queryStarted();
+            stats.queryStopped();
+            BasicQueryInfo queryInfo = dispatchQuery.getBasicQueryInfo();
+            stats.queuedQueryFailed(queryInfo.getQueryStats().getQueuedTime(), Optional.ofNullable(queryInfo.getErrorCode()));
+        }
+        finally {
+            // execution MUST be added to the expiration queue or there will be a leak
+            queryTracker.expireQuery(dispatchQuery.getQueryId());
+        }
+    }
+
+    public ListenableFuture<?> waitForDispatched(QueryId queryId)
+    {
+        return queryTracker.tryGetQuery(queryId)
+                .map(dispatchQuery -> {
+                    dispatchQuery.recordHeartbeat();
+                    return dispatchQuery.getDispatchedFuture();
+                })
+                .orElseGet(() -> immediateFuture(null));
+    }
+
+    public List<BasicQueryInfo> getQueries()
+    {
+        return queryTracker.getAllQueries().stream()
+                .map(DispatchQuery::getBasicQueryInfo)
+                .collect(toImmutableList());
+    }
+
+    public BasicQueryInfo getQueryInfo(QueryId queryId)
+    {
+        return queryTracker.getQuery(queryId).getBasicQueryInfo();
+    }
+
+    public Optional<DispatchInfo> getDispatchInfo(QueryId queryId)
+    {
+        return queryTracker.tryGetQuery(queryId)
+                .map(dispatchQuery -> {
+                    dispatchQuery.recordHeartbeat();
+                    return dispatchQuery.getDispatchInfo();
+                });
+    }
+
+    public void cancelQuery(QueryId queryId)
+    {
+        queryTracker.tryGetQuery(queryId)
+                .ifPresent(DispatchQuery::cancel);
+    }
+
+    private static class DispatchQueryCreationFuture
```

Comment:
Turns out it's a different semantic.  An attempt to cancel this future has no effect, whereas cancelling the returned future from `nonCancellationPropagating` may be cancelled (however this will not be propagated to the inner future).